src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,3315 @@
+
+    // map (block_height, vtxindex) to (non-prepare-ancestor-height, non-prepare-ancestor-vtxindex, total_burnt)
+    let mut ancestors = BTreeMap::new();
+
+    // map (non-prepare-ancestor-height, non-prepare-ancestor-vtxindex) to (set-of-block-heights, total_burnt)
+    // that contain descendants
+    let mut ancestor_confirmations: BTreeMap<(u64, u32), (HashSet<u64>, u64)> = BTreeMap::new();
+
+    // calculate each block-commit's parents
+    for prepare_block_ops in prepare_ops.iter() {
+        for opdata in prepare_block_ops.iter() {
+            parents.insert(
+                (opdata.block_height, opdata.vtxindex),
+                (
+                    opdata.burn_fee,
+                    opdata.parent_block_ptr as u64,
+                    opdata.parent_vtxindex as u32,
+                ),
+            );
+        }
+    }
+
+    // calculate the ancestor map -- find the highest ancestor for each prepare-phase block-commit
+    // that is _not_ in the prepare phase.
+    for prepare_block_ops in prepare_ops.iter().rev() {
+        for opdata in prepare_block_ops.iter() {
+            let mut cursor = (opdata.block_height, opdata.vtxindex);
+            let mut total_burnt = 0;
+            while !ancestors.contains_key(&cursor) {
+                if let Some((burnt, parent_block, parent_vtxindex)) = parents.get(&cursor) {
+                    cursor = (*parent_block, *parent_vtxindex);
+                    total_burnt += *burnt;
+                } else {
+                    break;
+                }
+            }
+            if !ancestors.contains_key(&cursor) {
+                ancestors.insert(
+                    (opdata.block_height, opdata.vtxindex),
+                    (cursor.0, cursor.1, total_burnt),
+                );
+            }
+        }
+    }
+
+    // calculate the ancestor confirmations -- figure out how many distinct blocks contain
+    // block-commits that descend from each pre-prepare-phase ancestor
+    for prepare_block_ops in prepare_ops.iter() {
+        for opdata in prepare_block_ops.iter() {
+            if let Some((ancestor_height, ancestor_vtxindex, total_burnt)) =
+                ancestors.get(&(opdata.block_height, opdata.vtxindex))
+            {
+                if let Some((ref mut confirmed_block_set, ref mut ancestor_burnt)) =
+                    ancestor_confirmations.get_mut(&(*ancestor_height, *ancestor_vtxindex))
+                {
+                    confirmed_block_set.insert(opdata.block_height);
+                    *ancestor_burnt = cmp::max(*total_burnt, *ancestor_burnt);
+                } else {
+                    let mut block_set = HashSet::new();
+                    block_set.insert(opdata.block_height);
+                    ancestor_confirmations.insert(
+                        (*ancestor_height, *ancestor_vtxindex),
+                        (block_set, *total_burnt),
+                    );
+                }
+            }
+        }
+    }
```

Comment:
I'm having a hard time figuring out how `total_burnt` and `confirmed_block_set` are being calculated. It looks to me like the following is what is happening:

* For each block commit in the prepare phase, create an entry in `parents`: `(block commit -> parent_commit, op.burn)`
* For each block commit in the prepare phase, scan backwards through its ancestry until one of the following is true:
      a. `ancestors` contains the current cursor
      b. the cursor is outside the prepare phase
If the scan stopped because of (b), then add a new `ancestors` entry with the current cursor, as well as a sum of each commit visited during the scan. If the scan stopped because of (a), do nothing.
* For each block commit, see if it has an entry in the `ancestors` map, and if so, count it as a confirmation using the block_set, and apply a `cmp::max` on any associated burn sums.

I'm getting tripped up in two places:

In the second step, I may be missing something, but it seems like this would undercount confirmations. For example, if the block commits are:

```
A <- B <- C <- D <- E
   |  prepare phase  |
```

Then, this code would iterate like `[E, D, C, B]` (A is not in prepare phase and would be the likely anchor block).

At iteration `E`, the ancestor construction loop would scan from E -> D -> C -> B and then set cursor to A, at which point it would break at line 579, inserting `A` into `ancestors` with the sum of the burns.

At iteration `D`, the ancestor construction loop would scan from D->C->B, set cursor to A, and then end the while loop without inserting anything new into `ancestors`. Similar things would happen when it processed C and B.

This would mean that `ancestors` only contains one entry (rather than 4).

The second place I'm getting tripped up is in the use of `cmp::max` -- it seems like this should be summing burn amounts, not applying a max. For example, if the block commits were:

```
A <- B <- C
 ^------------ D <- E
   |  prepare phase  |
```

We'd want to give A the sum of all those burns, not the max of the two forks.

Reply:
I think there's a bug here -- `if !ancestors.contains_key(&cursor) {` should be `if !ancestors.contains_key((opdata.block_height, opdata.vtxindex)) {` (line 582).  What we're trying to do here is map _each prepare-phase commit_ to its highest non-prepare ancestor, as well as the total BTC burnt for it by the chain that connects this prepare-phase commit to it.

In light of that, I think you're also right about `cmp::max` -- I think we should be summing burns.

Reply:
> I think there's a bug here -- if !ancestors.contains_key(&cursor) { should be if !ancestors.contains_key(&(opdata.block_height, opdata.vtxindex)) { (line 582).

Yes, I think that change would fix this loop, but I think that if statement would always return true -- each block commit is only visited once in the outer loop. So it'd be better just to _always_ write to `ancestors` (after all, every block commit should have an entry in that map).

> In light of that, I think you're also right about cmp::max not being right, but I don't think adding them gets the true BTC burnt.

It does if you just sum after doing the ancestor lookup. That is, this process would work like:

1. Create a map from block commits to their last ancestor before the prepare phase starts (this is their affirmed anchor block). Don't try to do any burn amount calculation at this point.
2. Iterate over all commits, and compute two things:
    a. burn total for each candidate anchor block. this is just summing into a map `candidate -> burn_amt`
    b. potential confirmation count for each candidate. this is incrementing a map `candidate -> confirm_count` every time a new commit affirms the candidate in a new burn block.

That would calculate `sum(ABCDED'E')` which is I think the desired amount.

Reply:
Fixed in 5bc69e5